src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/ServerSessionConnectionReaderMiddleware.cs

```diff
@@ -25,26 +28,36 @@ public ServerSessionConnectionReaderMiddleware(HandshakeDelegate next, IServiceS
 
         public async Task OnConnectedAsync(FramingConnection connection)
         {
-            if (!_transportSettingsCache.TryGetValue(connection.ServiceDispatcher, out ITransportFactorySettings settings))
+            ITransportFactorySettings settings;
+            if (_transportSettingsCache.Contains(connection.ServiceDispatcher))
             {
-                BindingElementCollection be = connection.ServiceDispatcher.Binding.CreateBindingElements();
-                TransportBindingElement tbe = be.Find<TransportBindingElement>();
-                settings = new NetFramingTransportSettings
+                settings = (ITransportFactorySettings)_transportSettingsCache[connection.ServiceDispatcher];
+            }
+            else
+            {
+                lock (_lock)
                 {
-                    CloseTimeout = connection.ServiceDispatcher.Binding.CloseTimeout,
-                    OpenTimeout = connection.ServiceDispatcher.Binding.OpenTimeout,
-                    ReceiveTimeout = connection.ServiceDispatcher.Binding.ReceiveTimeout,
-                    SendTimeout = connection.ServiceDispatcher.Binding.SendTimeout,
-                    ManualAddressing = tbe.ManualAddressing,
-                    BufferManager = connection.BufferManager,
-                    MaxReceivedMessageSize = tbe.MaxReceivedMessageSize,
-                    MessageEncoderFactory = connection.MessageEncoderFactory
-                };
+                    BindingElementCollection be = connection.ServiceDispatcher.Binding.CreateBindingElements();
+                    TransportBindingElement tbe = be.Find<TransportBindingElement>();
+                    settings = new NetFramingTransportSettings
+                    {
+                        CloseTimeout = connection.ServiceDispatcher.Binding.CloseTimeout,
+                        OpenTimeout = connection.ServiceDispatcher.Binding.OpenTimeout,
+                        ReceiveTimeout = connection.ServiceDispatcher.Binding.ReceiveTimeout,
+                        SendTimeout = connection.ServiceDispatcher.Binding.SendTimeout,
+                        ManualAddressing = tbe.ManualAddressing,
+                        BufferManager = connection.BufferManager,
+                        MaxReceivedMessageSize = tbe.MaxReceivedMessageSize,
+                        MessageEncoderFactory = connection.MessageEncoderFactory
+                    };
+                    _transportSettingsCache[connection.ServiceDispatcher] = settings;
+                }
             }
 
             var channel = new ServerFramingDuplexSessionChannel(connection, settings, false, _servicesScopeFactory.CreateScope().ServiceProvider);
             channel.ChannelDispatcher = await connection.ServiceDispatcher.CreateServiceChannelDispatcherAsync(channel);
             await channel.StartReceivingAsync();
+            await channel.CloseAsync();
```

Comment:
does this need to be explicit ? the input/outclose gets called as soon as the EOF is read if I am not wrong ?

Reply:
There's a change so that CloseAsync will Dispose the scope which will dispose of any created types within that scope. Currently the GC is handling it, which should be fine, but it means any finalizable objects get cleaned up on the finalization queue which is single threaded and blocks GC. This is just being a better citizen when it comes to cleanup.